packages/help-center/src/hooks/use-post-by-key.ts

```diff
@@ -20,10 +20,50 @@ function fetchForKey( postKey: ReturnType< typeof getPostKey > ) {
 		  } );
 }
 
-export function usePostByKey( postKey: ReturnType< typeof getPostKey > | null ) {
+function fetchForUrl( url: string ) {
+	return canAccessWpcomApis()
+		? wpcomRequest( {
+				path: `help/article?post_url=${ encodeURIComponent( url ) }`,
+				apiNamespace: 'wpcom/v2/',
+				apiVersion: '2',
+		  } )
+		: apiFetch( {
+				path: `/help-center/fetch-post?post_url=${ encodeURIComponent( url ) }`,
+		  } );
+}
+
+export function usePostByKey( postKey: ReturnType< typeof getPostKey > | string | null ) {
+	const searchingByUrl = typeof postKey === 'string';
+	const queryClient = useQueryClient();
+
 	return useQuery( {
-		queryKey: [ 'support-status', postKey ],
-		queryFn: () => postKey && fetchForKey( postKey ),
+		queryKey: [ 'support-article', postKey ],
+		queryFn: () => {
+			if ( searchingByUrl ) {
+				return fetchForUrl( postKey );
+			} else if ( postKey ) {
+				return fetchForKey( postKey );
+			}
+		},
+		select: ( data: any ) => {
+			if ( searchingByUrl ) {
+				// Set the cache also with the post key to prevent extra fetches.
+				queryClient.setQueryData(
+					[ 'support-article', getPostKey( data.site_ID, data.ID ) ],
+					data
+				);
+			}
+
+			return {
+				content: data.content,
+				title: data.title,
+				link: data.URL,
+				postId: data.ID,
+				blogId: data.site_ID,
+				slug: data.slug,
+				articleUrl: data.URL,
+			};
+		},
```

Comment:
My desire was to not mess with the current hook too much as to require updating it and possibly other places people are navigating to articles. With that being said it is a bit challenging to make a hook that accepts either the key or the url and stores it in a way thats not redundant ie post cache by url and cached by key. I also was really confused by what the need for `getPostKey()` and all the alternative stuff that seems to be tied into this.

I think in an ideal world we would scrap the need for post and blog id's and just use the URL, because it is much simpler and requires less work. But maybe that is too big of a jump to start with to eliminate it completely and adjust all the places calling the initialRoute to a post? Do you have any thoughts on how you would structure this @alshakero ?

Reply:
I think it's alright to switch to URLs only. This hook is used only in the Help Center. I built this hook a few weeks ago [here](https://github.com/Automattic/wp-calypso/pull/92008), don't worry about nuking it.